packages/tre/src/plugins/queues/gateway.ts

```diff
@@ -0,0 +1,250 @@
+import assert from "assert";
+import crypto from "crypto";
+import v8 from "v8";
+import { Colorize, bold, green, grey, red, reset, yellow } from "kleur/colors";
+import { z } from "zod";
+import { Log, Timers } from "../../shared";
+import { Storage } from "../../storage";
+import { CoreHeaders, structuredSerializableReducers } from "../../workers";
+import { DispatchFetch, QueueConsumer } from "../shared";
+
+const DEFAULT_BATCH_SIZE = 5;
+const DEFAULT_BATCH_TIMEOUT = 1; // second
+const DEFAULT_RETRIES = 2;
+
+// https://github.com/cloudflare/workerd/blob/01b87642f4eac932aa9074d7e5eec4fd3c90968a/src/workerd/io/outcome.capnp
+const Outcome = {
+  UNKNOWN: 0,
+  OK: 1,
+  EXCEPTION: 2,
+  EXCEEDED_CPU: 3,
+  KILL_SWITCH: 4,
+  DAEMON_DOWN: 5,
+  SCRIPT_NOT_FOUND: 6,
+  CANCELED: 7,
+  EXCEEDED_MEMORY: 8,
+} as const;
+const OutcomeSchema = z.nativeEnum(Outcome);
+
+const QueueResponseSchema = z.object({
+  outcome: OutcomeSchema,
+  retryAll: z.boolean(),
+  ackAll: z.boolean(),
+  explicitRetries: z.string().array(),
+  explicitAcks: z.string().array(),
+  time: z.number().optional(), // (added by Miniflare)
+});
+type QueueResponse = z.infer<typeof QueueResponseSchema>;
+const exceptionQueueResponse: QueueResponse = {
+  outcome: Outcome.EXCEPTION,
+  retryAll: false,
+  ackAll: false,
+  explicitRetries: [],
+  explicitAcks: [],
+};
+
+export class Message {
+  #failedAttempts = 0;
+
+  constructor(
+    readonly id: string,
+    readonly timestamp: number,
+    readonly serialisedBody: Buffer
+  ) {}
+
+  incrementFailedAttempts(): number {
+    return ++this.#failedAttempts;
+  }
+}
+
+function formatQueueResponse(
+  queueName: string,
+  acked: number,
+  total: number,
+  time?: number
+) {
+  let colour: Colorize;
+  if (acked === total) colour = green;
+  else if (acked > 0) colour = yellow;
+  else colour = red;
+
+  let message = `${bold("QUEUE")} ${queueName} ${colour(`${acked}/${total}`)}`;
+  if (time !== undefined) message += grey(` (${time}ms)`);
+  return reset(message);
+}
+
+interface PendingFlush {
+  immediate: boolean;
+  timeout: unknown;
+}
+
+export type QueueEnqueueOn = (
+  queueName: string,
+  consumer: QueueConsumer,
+  messages: (Message | Buffer)[]
+) => void;
+
+// `QueuesGateway` slightly misrepresents what this class does. Each queue will
+// get a single `QueuesGateway` instance (per `Miniflare` instance). This class
+// is responsible for receiving queued messages, batching them up, dispatching
+// them, and retrying any messages that failed.
+export class QueuesGateway {
+  readonly #queueUrl: URL;
+  readonly #messages: Message[] = [];
+
+  #pendingFlush?: PendingFlush;
+
+  constructor(
+    private readonly log: Log,
+    _storage: Storage,
+    private readonly timers: Timers,
+    private readonly queueName: string,
+    private readonly dispatchFetch: DispatchFetch
+  ) {
+    this.#queueUrl = new URL(`http://localhost/${queueName}`);
```

Comment:
Is this _actually_ `localhost:80`? If not, what about something like `http://dummy` for clarity?